static/posts/pattern-matching/post.md

```diff
@@ -0,0 +1,952 @@
+  struct { double x; double y; double t; }; // a parallelogram
+} Dimensions;
+
+// ...
+
+int main() {
+  size_t size = 3;
+  Shape shapes[size];
+  
+  // Define the shapes
+
+  // Area should be pi
+  shapes[0].type = CIRCLE;
+  shapes[0].dimensions.r = 1.0;
+
+  // Area should be 6.0
+  shapes[1].type = RECTANGLE;
+  shapes[1].dimensions.w = 2.0;
+  shapes[1].dimensions.h = 3.0;
+
+  // Area should be 6.0
+  shapes[2].type = triangle;
+  shapes[2].dimensions.a = 3.0;
+  shapes[2].dimensions.b = 4.0;
+  shapes[2].dimensions.c = 5.0;
+
+  // Area should be 6.0
+  shapes[3].type = PARALLELOGRAM;
+  shapes[3].dimensions.x = 2.0;
+  shapes[3].dimensions.y = 3.0;
+  shapes[3].dimensions.t = M_PI / 2;
+
+  double totalArea = 0;
+  for (size_t i = 0; i < size; i++) {
+    totalArea += shape_area(shapes[i]);
+  }
+
+  // Total area should be pi + 6.0 + 6.0 + 6.0 = 21.14...
+  printf("Total area: %f\n", totalArea);
+
+  return 0;
+}
+```
+
+We can compile and run our program:
+
+```text
+$ gcc shapes.c -o shapes -lm
+$ ./shapes
+Total area: 16.579133
+```
+
+Wait, that's not right... It should have printed 21.14... What's going on here?
+There weren't any compile errors, everything should be fine right?
+
+Wrong. The C compiler will compile this without complaint. It will even run the
+program for you without throwing an error or
+[segfaulting](https://en.wikipedia.org/wiki/Segmentation_fault)! This is the
+worst kind of bug. Aside from manual validation, there's no indication that
+there is anything wrong.
+
+The problem is that we forgot to include a branch to calculate the area of a
+parallelogram in `shape_area()`. The parallelogram is being treated as if it
+were a triangle.
+
+Let's modify the code to get a closer look:
+
+```c
+double shape_area(Shape shape) {
+  Dimensions ds = shape.dimensions;
+  if (shape.type == CIRCLE) {
+    return M_PI * ds.r * ds.r;
+  } else if (shape.type == RECTANGLE) {
+    return ds.w * ds.h;
+  } else {
+    printf("a: %f\n", ds.a);
+    printf("b: %f\n", ds.b);
+    printf("c: %f\n", ds.c);
+    double s = (ds.a + ds.b + ds.c) / 2;
+    return sqrt(s * (s - ds.a) * (s - ds.b) * (s - ds.c));
+  }
+}
+```
+
+And we can run the program again:
+
+```text
+$ gcc shapes.c -o shapes -lm
+$ ./shapes
+a: 3.000000
+b: 4.000000
+c: 5.000000
+a: 2.000000
+b: 3.000000
+c: 1.570796
+Total area: 16.579133
+```
+
+😮 The compiler is using our parallelogram dimensions as if they were triangle
+dimensions!
+
+The data for each object is stored in a contiguous memory block and when a
+dimension is read on a shape, the compiler just looks at the position in memory
+that the data should be in.
+
+In a large code base, this kind of problem can take ages to discover and hours
+to debug. If you're not careful and you don't test your code adequately, this
+problem could easily go unnoticed and be released, only to be discovered much
+later on.
+
+We could modify the function so that the final else branch exits the program if
```

Comment:
Is "final else branch" correctly worded? this sounds like computer lingo

Reply:
It is!